// log/index.js

var Farm = require('./farm.js');
var config = require('config');
var LogEntry = require('./entry.js');
var constants = require('const');
var extension = require('extension');
var farms = {
    unnamed: new Farm(config.logTimeoutInSeconds, 'unnamed')
};
farms[constants.loglevel.debug] = new Farm(config.logTimeoutInSeconds, 'debug');
farms[constants.loglevel.info] = new Farm(config.logTimeoutInSeconds, 'info');
farms[constants.loglevel.warn] = new Farm(config.logTimeoutInSeconds, 'warn');
farms[constants.loglevel.error] = new Farm(config.logTimeoutInSeconds, 'error');
farms[constants.loglevel.critical] = new Farm(config.logTimeoutInSeconds, 'critical');

var log = function (options) {
    var logentry = new LogEntry(options);
    var f = farms[logentry.level];
    f ? f.plant(logentry) : farms.unnamed.plant(logentry);
};

var refineLog = function (headers, body) {
    if (!body) {
        body = headers;
        headers = {};
    }
    else {
        headers = headers || {};
    }
    headers.insertedOn = extension.datetime.chinaNow();
    return { headers: headers, body: body };
};

var logFactory = function (level, cut) {
    return function (headers, body) {
        var logData = refineLog(headers, body);
        logData.level = level;
        log({ data: logData, cut: cut });
    };
};

module.exports = {
    critical: logFactory(constants.loglevel.critical),
    error: logFactory(constants.loglevel.error),
    warn: logFactory(constants.loglevel.warn),
    info: logFactory(constants.loglevel.info),
    debug: logFactory(constants.loglevel.debug, true),
};
