// smsvc/smsvc.js

var db = require('dbaccess');
var bluebird = require('bluebird');
var extension = require('extension');
var constants = require('const');
var imageUpload = require('imageUpload');
var errdef = require('errdef');
var resxutil = require('resxutil');
var config = require('config');
var Resource = require('resource');
var validate = require('validate');
var _ = require('underscore');
var infra = require('infra');
var smsproviders = require('smsproviders');
var log = require('log');

/*
{
smsId:,
provider:,
type:,
createdOn:,
userId:,
code:,
expiredOn:,
claimedOn:,
}
*/
var SmsVC = function (webreq) {
    Resource.call(this, webreq);
    this.vcType = webreq.search.type;
};

SmsVC.prototype = Object.create(Resource.prototype);

SmsVC.prototype._accessibility = function () {
    var httpOptions = new resxutil.HttpOptions();
    return bluebird.resolve(httpOptions.setStatus('POST', this.operator ? 200 : 401));
};

var handlers = {};
handlers[constants.smsvc.type.mobileV] = function () {
    var me = this, smsp = smsproviders.getSmsProvider();
    return db.user.getOneUserFieldsBy({ _id: me.operator.objId }, { mobileV: 1 })
    .then(function (targetUser) {
        if (!targetUser) //basically should not happen because this func needs authn
            throw new errdef.DeferChainKiller(extension.http.webres400());
        else if (targetUser.mobileV) //already have mobileV
            throw new errdef.DeferChainKiller(extension.http.webres409());
        else {
            return smsp.handler({});
        }
    })
    .catch(errdef.SmsError, function (err) {
        log.error({ type: 'sms error' }, { err: err.stack });
        throw new errdef.DeferChainKiller(extension.http.webres500());
    })
    .then(function (smsRet) {
        var now = new Date();
        var exp = extension.datetime.getDateWithOffset({ ref: now, minutes: 30 });
        return db.smsvc.insertOneSmsVC({
            smsId: smsRet.smsId,
            provider: smsp.name,
            type: constants.smsvc.type.mobileV,
            userId: me.operator.objId,
            code: smsRet.code,
            createdOn: now,
            expiredOn: exp
        });
    })
    .then(function (inserted) {
        //maybe i should move this http-faced code to upper layer, not the sms handler layer
        return extension.http.webres201(constants.mime.json, JSON.stringify({ id: inserted._id }));
    });
};

SmsVC.prototype._post = function () {
    var me = this, targetSite;
    return me._accessibility()
    .then(function (httpOptions) {
        var status = httpOptions.getStatus('POST');
        if (status.isOK()) {
            var handler = handlers[me.vcType];
            if (!handler)
                throw new errdef.DeferChainKiller(extension.http.webres400());
            return handler.call(me);
        }
        else
            throw new errdef.DeferChainKiller(status.toWebres());
    })
    .then(function (ret) {
        //just a placeholder for some post-send things
        return ret;
    })
    .catch(errdef.DeferChainKiller, function (err) {
        return err.attach;
    });
};

module.exports = SmsVC;
