// resources/index.js
exports.usersResFactory = require('users');
exports.userCredentialResFactory = require('usercredentials');
exports.sitesResFactory = require('sites');
exports.siteIdlesResFactory = require('siteidles');
exports.sessionsResFactory = require('sessions');
exports.membersResFactory = require('members');
exports.activitiesResFactory = require('activities');
exports.vendorsResFactory = require('vendors');
exports.usercontactsResFactory = require('usercontacts');
exports.vendorpreviewsResFactory = require('vendorpreview');
exports.messagesResFactory = require('messages');
exports.notesResFactory = require('notes');
exports.availabilityResFactory = require('availability');
exports.footprintsResFactory = require('footprints');
exports.wxResFactory = require('wx');
exports.userfansResFactory = require('userfans');
exports.sitefansResFactory = require('sitefans');
exports.actcensusResFactory = require('actcensus');
exports.pwdresetResFactory = require('pwdreset');
exports.pendingNotesResFactory = require('pendingnotes');
exports.creditsResFactory = require('credits');
exports.smsvcResFactory = require('smsvc');
