// resxutil/user.js

var db = require('dbaccess');
var bluebird = require('bluebird');

exports.getFields = { '_id': 1, 'personal': 1, 'email': 1, 'headUrl': 1, 'fans': 1, 'subscribe': 1, 'mobileV': 1, 'special': 1 };
exports.convUserInfo = function (user) {
    return { 'id': user._id.toString(), 'personal': user.personal, 'email': user.email, 'headUrl': user.headUrl, 'fansCount': user.fans.length, 'subscribe': { usersCount: user.subscribe.users.length, sitesCount: user.subscribe.sites.length }, 'mobileV': user.mobileV && user.personal && user.personal.phone && user.mobileV.phone == user.personal.phone, special: user.special };
};
exports.checkDup = function (email, nickName, selfObjId) {
    var orFilter = [];
    if (email) orFilter.push({ email: email });
    if (nickName) orFilter.push({ 'personal.nickName': nickName });
    if (orFilter.length > 0) {
        var filter = { '$or': orFilter };
        if (selfObjId) filter._id = { $ne: selfObjId };
        return db.user.getOneUserFieldsBy(filter, { '_id': 1 })
        .then(function (dupUser) {
            return !Boolean(dupUser);
        });
    }
    else
        return bluebird.resolve(true);
};
exports.getRelatedUserFields = { '_id': 1, 'personal': 1, 'headUrl': 1, 'fans': 1, 'mobileV': 1 };
exports.convRelatedUserInfo = function (user) {
    return { 'id': user._id.toString(), 'personal': user.personal, 'headUrl': user.headUrl, 'fansCount': user.fans.length, 'mobileV': user.mobileV && user.personal && user.personal.phone && user.mobileV.phone == user.personal.phone };
};
exports.userExistence = function (uniqueUserIdList) {
    if (!uniqueUserIdList || uniqueUserIdList.length == 0)
        return bluebird.resolve(true);
    else {
        return db.user.getUsersFieldsBy({ _id: { $in: uniqueUserIdList } }, { _id: 1 })
        .then(function (users) {
            return users.length === uniqueUserIdList.length;
        });
    }
};
exports.getPublicUserFields = { '_id': 1, 'personal': 1, 'headUrl': 1, 'fans': 1, 'mobileV': 1 };
exports.convPublicUserInfo = function (user) {
    return { 'id': user._id.toString(), 'personal': { contact: user.personal.contact, gender: user.personal.gender, nickName: user.personal.nickName }, 'headUrl': user.headUrl, 'fansCount': user.fans.length, 'mobileV': user.mobileV && user.personal && user.personal.phone && user.mobileV.phone == user.personal.phone };
};
