// resxutil/user.js

var db = require('dbaccess');
var bluebird = require('bluebird');

exports.getFields = { '_id': 1, 'personal': 1, 'email': 1, 'headUrl': 1 };
exports.convUserInfo = function (user) {
    return { 'id': user._id.toString(), 'personal': user.personal, 'email': user.email, 'headUrl': user.headUrl };
};
exports.checkDup = function (email, nickName, selfObjId) {
    var orFilter = [];
    if (email) orFilter.push({ email: email });
    if (nickName) orFilter.push({ 'personal.nickName': nickName });
    if (orFilter.length > 0) {
        var filter = { '$or': orFilter };
        if (selfObjId) filter._id = { $ne: selfObjId };
        return db.user.getOneUserFieldsBy(filter, { '_id': 1 })
        .then(function (dupUser) {
            return !Boolean(dupUser);
        });
    }
    else
        return bluebird.resolve(true);
};
exports.getRelatedUserFields = { '_id': 1, 'personal': 1, 'headUrl': 1 };
exports.convRelatedUserInfo = function (user) {
    return { 'id': user._id.toString(), 'personal': user.personal, 'headUrl': user.headUrl };
};
exports.userExistence = function (uniqueUserIdList) {
    if (!uniqueUserIdList || uniqueUserIdList.length == 0)
        return bluebird.resolve(true);
    else {
        return db.user.getUsersFieldsBy({ _id: { $in: uniqueUserIdList } }, { _id: 1 })
        .then(function (users) {
            return users.length === uniqueUserIdList.length;
        });
    }
};
exports.getPublicUserFields = { '_id': 1, 'personal.contact': 1, 'personal.gender': 1, 'personal.nickName': 1, 'headUrl': 1 };
exports.convPublicUserInfo = function (user) {
    return { 'id': user._id.toString(), 'personal': user.personal, 'headUrl': user.headUrl };
};
