// errdef/index.js

//BadArgumentsError
var BadArgumentsError = function (msg) {
    this._stack = (new Error()).stack;
    this.name = 'BadArgumentsError';
    this.message = msg;
};
BadArgumentsError.prototype = Object.create(Error.prototype);
Object.defineProperty(BadArgumentsError.prototype, 'stack', {
    get: function () {
        return this._stack;
    }
});

//ImageUploadError
var ImageUploadError = function (msg) {
    this._stack = (new Error()).stack;
    this.name = 'ImageUploadError';
    this.message = msg;
};
ImageUploadError.prototype = Object.create(Error.prototype);
Object.defineProperty(ImageUploadError.prototype, 'stack', {
    get: function () {
        return this._stack;
    }
});

//DeferChainKiller
var DeferChainKiller = function (attach) {
    this.attach = attach;
};
DeferChainKiller.prototype = Object.create(Error.prototype);

//EmailError
var EmailError = function (inner) {
    this._stack = (new Error()).stack;
    this.name = 'EmailError';
    this.inner = inner;
};
EmailError.prototype = Object.create(Error.prototype);
Object.defineProperty(EmailError.prototype, 'stack', {
    get: function () {
        return this._stack;
    }
});

//SmsError
var SmsError = function (inner) {
    this._stack = (new Error()).stack;
    this.name = 'SmsError';
    this.inner = inner;
};
SmsError.prototype = Object.create(Error.prototype);
Object.defineProperty(SmsError.prototype, 'stack', {
    get: function () {
        return this._stack;
    }
});

exports.BadArgumentsError = BadArgumentsError;
exports.ImageUploadError = ImageUploadError;
exports.DeferChainKiller = DeferChainKiller;
exports.EmailError = EmailError;
exports.SmsError = SmsError;
